### PR TITLE
add C++ Server

### PR DESCRIPTION
**Pull request by pingfengluo**
- [x]  add C++ Server using WebSocket

- [x] add C++ Client using WebSocket

- [x] support auto detect and recycle the socket connection resource that not available for a long time

- [x]  a CI test to WebSocket Server

### CODE REVIEW

File: sherpa/cpp_api/websocket/client.cc [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+        beast::flat_buffer buffer;
+        ws_.read(buffer);
+        std::string message = beast::buffers_to_string(buffer.data());
+        SHERPA_LOG(INFO) << message;
+        json::object obj = json::parse(message).as_object();
+        if (obj["status"] != "ok") {
+          break;
+        }
+      }
+    } catch (beast::system_error const& se) {
```
Comment (csukuangfj) [inline sherpa/cpp_api/websocket/client.cc:64]:
```suggestion
    } catch (const beast::system_error& se) {
```

To be consistent with other code in sherpa, we use
```
const some_type& var;
```

File: sherpa/cpp_api/websocket/client.cc [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+      data.push_back(static_cast<int16_t>(wave_data[j].item<float>() * 32768));
+    }
+    // send PCM data with 16k1c16b format
+    client.SendBinaryData(data.data(), data.size() * sizeof(int16_t));
+    SHERPA_LOG(INFO) << "Send " << data.size() << " samples";
+    std::this_thread::sleep_for(
+        std::chrono::milliseconds(static_cast<int>(interval * 1000 * 0.8)));
+  }
+  SHERPA_LOG(INFO) << "No more data";
+  client.Join();
```
Comment (csukuangfj) [inline sherpa/cpp_api/websocket/client.cc:135]: I suggest that you invoke `Join()` inside the destructor of the client.
```cpp

~WebSocketClient() {
  Join();
}
```

And make `Join()` a private method.

Comment (pingfengluo) [reply sherpa/cpp_api/websocket/client.cc]: done. thank you!

**Comment by pingfengluo** [comment]
close now, should make another PR

**Pull request closed without merging**